core/src/main/scala/org/bitcoins/core/wallet/utxo/InputInfo.scala

```diff
@@ -0,0 +1,446 @@
+package org.bitcoins.core.wallet.utxo
+
+import org.bitcoins.core.currency.CurrencyUnit
+import org.bitcoins.core.protocol.script.{
+  ConditionalScriptPubKey,
+  EmptyScriptPubKey,
+  EmptyScriptWitness,
+  LockTimeScriptPubKey,
+  MultiSignatureScriptPubKey,
+  NonStandardScriptPubKey,
+  P2PKHScriptPubKey,
+  P2PKScriptPubKey,
+  P2PKWithTimeoutScriptPubKey,
+  P2SHScriptPubKey,
+  P2WPKHWitnessSPKV0,
+  P2WPKHWitnessV0,
+  P2WSHWitnessSPKV0,
+  P2WSHWitnessV0,
+  RawScriptPubKey,
+  ScriptPubKey,
+  ScriptWitness,
+  ScriptWitnessV0,
+  UnassignedWitnessScriptPubKey,
+  WitnessCommitment,
+  WitnessScriptPubKey,
+  WitnessScriptPubKeyV0
+}
+import org.bitcoins.core.protocol.transaction.{
+  TransactionOutPoint,
+  TransactionOutput
+}
+import org.bitcoins.core.script.crypto.HashType
+import org.bitcoins.crypto.{ECPublicKey, Sign}
+
+sealed trait InputInfo {
```

Comment:
So I think part of the reason we keep having design issues is this trait where we bundle in everything as Options. I think we need to break this into smaller pieces so that we can then make this a proper ADT without all `Options` (this many options in one trait usually indicates a design flaw). Here some pseudocode of what I was thinkikng: 

```scala
//this is needed for every utxo we spend, so we make this it's own case class
//and use a has-a relationship rather than a "is-a" relationship
case class Output(outPoint: TransactionOutPoint, output: TransactionOutput) {
  def scriptPubKey: ScriptPubKey = output.scriptPubKey

  def amount: CurrencyUnit = output.value
}

//here is where we start getting into ADTs
//for encoding things we need to "satisfy" the script
sealed trait InputInfo {
  def output: Output
}

case class BaseP2SHInputInfo(redeemScript: RawScriptPubKey,
                             override val output: Output) extends InputInfo

case class RawWitSPKInputInfo(witnessScript: WitnessScriptPubKey,
                              override val output: Output) extends InputInfo

case class P2PKHInputInfo(publicKey: ECPublicKey, override val output: Output) extends InputInfo

case class P2SHWitSPKInputInfo(redeemScript: WitnessScriptPubKey, scriptWitness: ScriptWitness, output: Output) extends InputInfo

//....
```

The unfortunately thing with this is `Output` doesn't give you information about the `ScriptPubKey` nested inside of it